### PR TITLE
Lifecycle transitions for LifecycleComponent

### PR DESCRIPTION
**Pull request by domire8**
WIP

### CODE REVIEW

File: source/modulo_components/include/modulo_components/LifecycleComponent.hpp [hunk 1]
```diff
@@ -38,10 +67,65 @@ class LifecycleComponent : public ComponentInterface<rclcpp_lifecycle::Lifecycle
...
+  /**
+    * @brief Transition callback for state 'Deactivating'.
+    * @details on_deactivate callback is being called when the lifecycle node enters the 'Deactivating' state.
+    * Depending on the return value of this function, the node may either transition to the 'Inactive' state via the
+    * 'deactivate' transition or goes to 'ErrorProcessing'.\n
+    * TRANSITION_CALLBACK_SUCCESS transitions to "Inactive"\n
+    * TRANSITION_CALLBACK_FAILURE, TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'\n
+    */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_shutdown(const rclcpp_lifecycle::State& state) override;
```
Comment (eeberhard) [inline source/modulo_components/include/modulo_components/LifecycleComponent.hpp:124]:
```suggestion
  /**
    * @brief Transition callback for state 'ShuttingDown'.
    * @details on_shutdown callback is being called when the lifecycle node enters the 'ShuttingDown' state.
    * This transition can be called from the 'Unconfigured', 'Inactive' and 'Active' states.
    * Depending on the return value of this function, the node may either transition to the 'Finalized' state via the
    * 'shutdown' transition or goes to 'ErrorProcessing'.\n
    * TRANSITION_CALLBACK_SUCCESS transitions to "Finalized"\n
    * TRANSITION_CALLBACK_FAILURE, TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'\n
    */
  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
  on_shutdown(const rclcpp_lifecycle::State& state) override;
```

File: source/modulo_components/include/modulo_components/LifecycleComponent.hpp [hunk 2]
```diff
@@ -38,10 +67,65 @@ class LifecycleComponent : public ComponentInterface<rclcpp_lifecycle::Lifecycle
   void add_output(const std::string& signal_name, const std::shared_ptr<DataT>& data, bool fixed_topic = false);
 
 private:
+  /**
+   * @brief Transition callback for state 'Configuring'.
+   * @details on_configure callback is being called when the lifecycle node enters the 'Configuring' state.
+   * Depending on the return value of this function, the node may either transition to the 'Inactive' state via the
+   * 'configure' transition or stays 'Unconfigured.\n
```
Comment (eeberhard) [inline source/modulo_components/include/modulo_components/LifecycleComponent.hpp:74]:
```suggestion
   * 'configure' transition, stay 'Unconfigured' or go to 'ErrorProcessing'.\n
```

File: source/modulo_components/include/modulo_components/LifecycleComponent.hpp [hunk 3]
```diff
@@ -38,10 +67,65 @@ class LifecycleComponent : public ComponentInterface<rclcpp_lifecycle::Lifecycle
...
+    * TRANSITION_CALLBACK_FAILURE, TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'\n
+    */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+  on_cleanup(const rclcpp_lifecycle::State&) override;
+
+  /**
+  * @brief Transition callback for state 'Activating'.
+  * @details on_activate callback is being called when the lifecycle node enters the 'Activating' state.
+  * Depending on the return value of this function, the node may either transition to the 'Active' state via the
+  * 'activate' transition or goes to 'ErrorProcessing'.\n
```
Comment (eeberhard) [inline source/modulo_components/include/modulo_components/LifecycleComponent.hpp:97]:
```suggestion
  * 'activate' transition, stay 'Inactive' or go to 'ErrorProcessing'.\n
```

**Review by eeberhard — COMMENTED** [review]
Some initial comments to revise the transition docstrings, as this helps keep a good overview of what each function should do

**Review by eeberhard — APPROVED** [review]

**Review by domire8 — COMMENTED** [review]
Nice thanks a lot, let's merge and then continue from there as discussed. The private inheritance has to go as it creates too many issues.

File: source/modulo_components/src/LifecycleComponent.cpp [hunk 4]
```diff
@@ -1,18 +1,222 @@
...
+}
 
 void LifecycleComponent::step() {
-  // TODO do this only if active and add on_step
-//  this->publish_predicates();
-//  this->publish_outputs();
-//  this->evaluate_daemon_callbacks();
-//  this->on_step();
+  if (this->get_predicate("is_active")) {
+    this->publish_predicates();
```
Comment (domire8) [inline source/modulo_components/src/LifecycleComponent.cpp:19]:
```suggestion
  this->publish_predicates();
  if (this->get_predicate("is_active")) {
```

**Pull request merged**